internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,278 @@
+	Kubeclient kubernetes.Interface
+
+	// GRPCPort is the container grpc port which is defaulted to 50051
+	GRPCPort int32
+}
+
+// Create returns a bundle registry pod built from an index image after verifying successful creation
+func (rp *RegistryPod) Create(ctx context.Context) (*corev1.Pod, string, error) {
+	pod, err := rp.create(ctx)
+	if err != nil {
+		podLogs, logErr := rp.FetchPodLogs(ctx, pod)
+		if logErr != nil {
+			return nil, podLogs, fmt.Errorf("error creating pod: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error creating pod: %v", err)
+	}
+
+	err = rp.verifyPodRunning(ctx, pod)
+	if err != nil {
+		podLogs, logErr := rp.FetchPodLogs(ctx, pod)
+		if logErr != nil {
+			return nil, podLogs, fmt.Errorf("error verifying pod creation: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error verifying pod creation: %v", err)
+	}
+
+	return pod, "", nil
+}
+
+// create returns a pod built from an index image or returns if an existing pod with the same name is found
+func (rp *RegistryPod) create(ctx context.Context) (*corev1.Pod, error) {
+	// call init() to set defaults and validate the RegistryPod struct
+	err := rp.init()
+	if err != nil {
+		return nil, fmt.Errorf("error in initializing and validating registry pod: %v", err)
+	}
+
+	// call build() to make the pod definition
+	pod, err := rp.build()
+	if err != nil {
+		return nil, fmt.Errorf("error in building registry pod: %v", err)
+	}
+
+	// Check if Pod already exists
+	tmpPod, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Create pod in kubernetes cluster using the above pod definition and clientset
+			pod, err = rp.Kubeclient.CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("error creating registry pod: %v", err)
+			}
+		} else {
+			return nil, fmt.Errorf("error getting existing registry pod: %v", err)
+		}
+	} else {
+		return tmpPod, nil
+	}
+
+	return pod, nil
+}
+
+func (rp *RegistryPod) verifyPodRunning(ctx context.Context, pod *corev1.Pod) error {
+	// Upon creation of new pod, poll and verify that pod status is running
+	podCheck := wait.ConditionFunc(func() (done bool, err error) {
+		p, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("error getting pod %s: %w", pod.Name, err)
+		}
+		return p.Status.Phase == corev1.PodRunning, nil
+	})
+
+	err := wait.PollImmediateUntil(time.Duration(200*time.Millisecond), podCheck, ctx.Done())
+	if err != nil {
+		return fmt.Errorf("error waiting for registry pod %s to run: %v", pod.Name, err)
+	}
+
+	return err
+}
+
+func (rp *RegistryPod) init() error {
+	rp.setDefaults()
+	err := rp.validate()
+	if err != nil {
+		return fmt.Errorf("error in validating registry pod: %v", err)
+	}
+	return nil
+}
```

Comment:
```suggestion
```
I do not think that we should have RegistryPod.Init. when we initialize a struture/object we have a func called new

Reply:
Having an `init` or `initialize` method on a struct is a common Go idiom.

Reply:
In order to simplify, the init is more generic since we are speaking to init a process, file or action and it can indeed do not return a value such as a main(). The newObject is specific to initialized a structure what matches with my suggestion here in the cleanups.   

Just to clarifies the difference in both usages following some examples extracted from the controller-runtime.

**Init usage**

```go
func init() {
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		panic("Could not determine the path of the BinPathFinder")
	}
	assetsPath = filepath.Join(filepath.Dir(thisFile), "..", "assets", "bin")
}
```
From: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/testing/integration/internal/bin_path_finder.go#L13

```go
func init() {
	RuntimeLog = log.Log.WithName("controller-runtime")
}

```
From: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/log/log.go

**New usage**

```go
// CertWatcher watches certificate and key files for changes.  When either file
// changes, it reads and parses both and calls an optional callback with the new
// certificate.
type CertWatcher struct {
	sync.Mutex

	currentCert *tls.Certificate
	watcher     *fsnotify.Watcher

	certPath string
	keyPath  string
}

// New returns a new CertWatcher watching the given certificate and key.
func New(certPath, keyPath string) (*CertWatcher, error) {
	var err error

	cw := &CertWatcher{
		certPath: certPath,
		keyPath:  keyPath,
	}

	// Initial read of certificate and key.
	if err := cw.ReadCertificate(); err != nil {
		return nil, err
	}

	cw.watcher, err = fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}

	return cw, nil
}

```

```
// NewInformersMap creates a new InformersMap that can create informers for
// both structured and unstructured objects.
func NewInformersMap(config *rest.Config,
	scheme *runtime.Scheme,
	mapper meta.RESTMapper,
	resync time.Duration,
	namespace string) *InformersMap {

	return &InformersMap{
		structured:   newStructuredInformersMap(config, scheme, mapper, resync, namespace),
		unstructured: newUnstructuredInformersMap(config, scheme, mapper, resync, namespace),

		Scheme: scheme,
	}
}
```

Just to clarifies the difference in both usages following Golang docs: 

**init**: https://golang.org/doc/effective_go.html#init


> Finally, each source file can define its own niladic init function to set up whatever state is required. (Actually each file can have multiple init functions.) And finally means finally: init is called after all the variable declarations in the package have evaluated their initializers, and those are evaluated only after all the imported packages have been initialized.

> Besides initializations that cannot be expressed as declarations, a common use of init functions is to verify or repair 
correctness of the program state before real execution begins.

```
func init() {
    if user == "" {
        log.Fatal("$USER not set")
    }
    if home == "" {
        home = "/home/" + user
    }
    if gopath == "" {
        gopath = home + "/go"
    }
    // gopath may be overridden by --gopath flag on command line.
    flag.StringVar(&gopath, "gopath", gopath, "override default GOPATH")
}
```

**Structures:** https://golang.org/doc/effective_go.html#composite_literals

> Constructors and composite literals
> Sometimes the zero value isn't good enough and an initializing constructor is necessary, as in this example derived from package os.

```
func NewFile(fd int, name string) *File {
    if fd < 0 {
        return nil
    }
    f := new(File)
    f.fd = fd
    f.name = name
    f.dirinfo = nil
    f.nepipe = 0
    return f
}
```

> There's a lot of boiler plate in there. We can simplify it using a composite literal, which is an expression that creates a new instance each time it is evaluated.

```go
func NewFile(fd int, name string) *File {
    if fd < 0 {
        return nil
    }
    f := File{fd, name, nil, 0}
    return &f
}
```

Reply:
This is an internal method on a struct, not a global `init` function. Furthermore this is just a wrapper for initialization code that can be used in `NewRegistryPod`, so if you want to debate the name that's fine but saying "don't do this" is wrong.